dvc/stage/cache.py

```diff
@@ -170,12 +166,12 @@ def save(self, stage):
         COMPILED_LOCK_FILE_STAGE_SCHEMA(cache)
 
         path = PathInfo(self._get_cache_path(cache_key, cache_value))
-        self.tree.makedirs(path.parent)
+        self.repo.cache.local.makedirs(path.parent)
         tmp = tempfile.NamedTemporaryFile(delete=False, dir=path.parent).name
         assert os.path.exists(path.parent)
         assert os.path.isdir(path.parent)
         dump_yaml(tmp, cache)
-        self.tree.move(PathInfo(tmp), path)
+        self.repo.cache.local.move(PathInfo(tmp), path)
```

Comment:
Logic related to saving stage cache object, should really be in the cache itself.